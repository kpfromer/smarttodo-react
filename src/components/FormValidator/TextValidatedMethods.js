import validateIsEmail from "validator/lib/isEmail";

export const required = isRequired => ({
  hint: name => isRequired ? `${name} is required` : '',
  validate: value => isRequired ? !!value : true
});

export const minLength = minValue => ({
  hint: name => `${name} must be ${minValue + 1} characters or more`,
  validate: value => value.length > minValue
});

export const maxLength = maxValue => ({
  hint: name => `${name} must be ${maxValue - 1} characters or less`,
  validate: value => value.length < maxValue
});

export const email = isEmail => ({
  hint: () => isEmail ? `Must be a valid email` : '',
  validate: value => isEmail ? validateIsEmail(value) : true
});