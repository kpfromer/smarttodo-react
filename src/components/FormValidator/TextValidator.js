import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { isArray, isPlainObject } from 'lodash';

import { ValidatorContext } from './ValidatorContext';
import TextField from '@material-ui/core/TextField';

const validateShape = PropTypes.shape({
  text: PropTypes.string.isRequired,
  validate: PropTypes.func.isRequired
}).isRequired;

export class TextValidator extends Component {

  static propTypes = {
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    validate: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.arrayOf(validateShape),
      validateShape
    ]).isRequired,
    validateOnBlur: PropTypes.bool,
    textFieldProps: PropTypes.object,
    name: PropTypes.string,
    updateFormInput: PropTypes.func,
    initializeInput: PropTypes.func,
    removeInput: PropTypes.func,
  };

  static defaultProps = {
    validateOnBlur: false
  };

  state = {
    error: false,
    errorText: ''
  };

  validateValue = value => {
    const { updateFormInput, validate, name } = this.props;

    let error = false;
    let errorText = '';

    const validatationObject = validateObj => {
      const isError = !validateObj.validate(value);
      if (isError) {
        error = true;
        errorText = validateObj.text;
      }
    };

    if (isArray(validate)) {
      for (const validateObj of validate) {
        validatationObject(validateObj);
        if (error) {
          break;
        }
      }
    } else if (isPlainObject(validate)) {
      validatationObject(validate);
    } else {
      error = !validate(value);
    }

    if (updateFormInput) {
      updateFormInput({
        name,
        value,
        error
      });
    }
    this.setState({ error, errorText });

    return { error };
  };

  componentDidMount() {
    const { initializeInput, name, value } = this.props;
    if (initializeInput) {
      initializeInput(name, value, this.validateValue);
    }
  }

  componentWillUnmount() {
    const { removeInput, name } = this.props;
    if (removeInput) {
      removeInput(name);
    }
  }

  componentDidUpdate(prevProps) {
    const { value: oldValue, name: oldName } = prevProps;
    const { value: newValue, name: newName, removeInput, initializeInput } = this.props;

    if (oldValue !== newValue) {
      this.validateValue(newValue);
    }

    if (oldName !== newName) {
      removeInput(oldName);
      initializeInput(newName);
    }
  }

  handleChange = event => {
    const { value } = event.target;
    const { error } = this.validateValue(value);
    this.props.onChange({ value, error });
  };

  handleBlur = event => {
    if (this.props.validateOnBlur) {
      this.validateValue(event.target.value);
    }
  };

  render() {
    const { value, textFieldProps } = this.props;
    const { error, errorText } = this.state;
    return (
      <TextField
        {...textFieldProps}
        onBlur={this.handleBlur}
        value={value}
        error={error}
        onChange={this.handleChange}
        helperText={errorText}
      />
    );
  }
}

export default props =>
  <ValidatorContext.Consumer>
    {
      ({ updateFormInput, initializeInput, removeInput }) =>
        <TextValidator
          {...props}
          updateFormInput={updateFormInput}
          initializeInput={initializeInput}
          removeInput={removeInput}
        />
    }
  </ValidatorContext.Consumer>
