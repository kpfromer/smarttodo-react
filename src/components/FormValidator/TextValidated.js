import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { email, maxLength, minLength, required } from './TextValidatedMethods';
import TextValidator from "./TextValidator";

export default class TextValidated extends Component {

  static propTypes = {
    hintName: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    name: PropTypes.string,
    textValidatorProps: PropTypes.object,
    required: PropTypes.bool,
    minLength: PropTypes.number,
    maxLength: PropTypes.number,
    email: PropTypes.bool
  };

  static defaultProps = {
    required: null,
    minLength: null,
    maxLength: null,
    email: null
  };

  validatedMethods = {
    required,
    minLength,
    maxLength,
    email
  };

  render() {
    const { name, value, onChange, hintName, textValidatorProps, ...validationTypes } = this.props;
    const validate = Object.keys(validationTypes)
      .filter(validationName => this.props[validationName] !== null)
      .map(validationName => {
        if (!Object.keys(this.validatedMethods).includes(validationName)) {
          throw new Error(`Invalid property of TextValidated: ${validationName}`);
        }

        const validateValue = this.props[validationName];

        if (!validateValue) {
          throw new Error(`Invalid property of TextValidated: ${validationName}`);
        }

        const { hint, validate } = this.validatedMethods[validationName](validateValue);
        return {
          text: hint(hintName),
          validate
      }
    });

    return (
      <TextValidator
        {...textValidatorProps}
        name={name}
        value={value}
        onChange={onChange}
        validate={validate}
      />
    );
  }
}
