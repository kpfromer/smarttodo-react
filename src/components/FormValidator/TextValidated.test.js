import * as validatedMethods from './TextValidatedMethods';
import TextValidated from "./TextValidated";
import { SetupComponent } from "react-component-setup";
import TextValidator from "./TextValidator";

jest.mock('./TextValidatedMethods');

const { shallow: setup } = SetupComponent({
  component: TextValidated,
  defaultProps: {
    hintName: 'exHintName',
    value: 'exValue',
    onChange: () => {}
  }
});

describe('TextValidated', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders without crashing', () => {
    setup();
  });
  it('renders a TextValidator', () => {
    const mockHandleChange = jest.fn();
    const { wrapper } = setup({
      name: 'exName',
      onChange: mockHandleChange
    });

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.props().onChange).toBe(mockHandleChange);
  }); // TODO: test that onChange func toBe (NOT TO EQUAL) provided func

  describe('validation', () => {
    it('provides TextValidated with validation based on validatedMethods', () => {
      const mockHint = jest.fn().mockReturnValue('length less than five!');
      const mockValidate = jest.fn();
      validatedMethods.maxLength.mockReturnValue({
        hint: mockHint,
        validate: mockValidate
      });

      const { wrapper } = setup({
        maxLength: 5
      });

      expect(wrapper.find(TextValidator).props().validate).toMatchSnapshot();
      expect(wrapper.find(TextValidator).props().validate[0].validate).toBe(mockValidate);

      expect(validatedMethods.maxLength).toHaveBeenCalledWith(5);
      expect(mockHint).toHaveBeenCalledWith('exHintName');
    });
  });

  describe('given `textValidatorProps`', () => {
    it('passes the values to the TextValidator', () => {
      const values = {
        margin: 'different',
        textFieldProps: {
          width: 'a value'
        }
      };
      const { wrapper } = setup({
        textValidatorProps: values
      });

      expect(wrapper.find(TextValidator).props()).toMatchObject(values);
    });
  });
});