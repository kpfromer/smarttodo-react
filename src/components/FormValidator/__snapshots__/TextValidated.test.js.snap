// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextValidated renders a TextValidator 1`] = `
<_default
  name="exName"
  onChange={[MockFunction]}
  validate={Array []}
  value="exValue"
/>
`;

exports[`TextValidated validation provides TextValidated with validation based on validatedMethods 1`] = `
Array [
  Object {
    "text": "length less than five!",
    "validate": [MockFunction],
  },
]
`;
